engines/syguspdr.cpp

```diff
@@ -0,0 +1,1542 @@
+          collateral = ic3formula_negate(collateral);
+        }
+
+        size_t idx = find_highest_frame(pg->idx, collateral);
+        assert(idx >= pg->idx);
+
+        assert(collateral.disjunction);
+        assert(collateral.term);
+        assert(collateral.children.size());
+        SygusPdr::constrain_frame(idx, collateral);
+
+        // re-add the proof goal at a higher frame if not blocked
+        // up to the frontier
+        if (idx < frontier_idx()) {
+          assert(!pg->target.disjunction);
+          proof_goals.new_proof_goal(pg->target, idx + 1, pg->next);
+        }
+
+      } else {
+        // could not block this proof goal
+        assert(collateral.term);
+        if (collateral.term == solver_true_)
+          ; // do nothing // because we need to extend it to 0 to 
+        proof_goals.new_proof_goal(collateral, pg->idx - 1, pg);
+      }
+    }  // end while(!proof_goals.empty())
+
+    assert(!(goal = IC3Formula()).term);  // in debug mode, reset it
+  }                                       // end while(reaches_bad(goal))
+
+  assert(proof_goals.empty());
+  return true;
+} // block_all
+
+// -----------------------------------------------------------------
+// allow input in the constraint (when there are additional constraints)
+// -----------------------------------------------------------------
+
+
+static bool subsumes(const IC3Formula &a, const IC3Formula &b)
+{
+  assert(a.disjunction);
+  assert(a.disjunction == b.disjunction);
+  const TermVec &ac = a.children;
+  const TermVec &bc = b.children;
+  // NOTE: IC3Formula children are sorted on construction
+  //       Uses unique id of term, from term->get_id()
+  return ac.size() <= bc.size()
+         && std::includes(bc.begin(), bc.end(), ac.begin(), ac.end());
+}
+
+
+void SygusPdr::constrain_frame(size_t i, const IC3Formula & constraint,
+                              bool new_constraint)
+{
+  assert(solver_context_ == 0);
+  assert(i < frame_labels_.size());
+  assert(constraint.disjunction);
+  if (has_assumptions || options_.sygus_use_operator_abstraction_ == 1) {
+    // options_.sygus_use_operator_abstraction_ == 2 will not introduce new inputs
+    // that must be kept
+    assert(ts_.no_next(constraint.term));
+  } else
+    assert(ts_.only_curr(constraint.term));
+
+  if (new_constraint) {
+    for (size_t j = 1; j <= i; ++j) {
+      std::vector<IC3Formula> & Fj = frames_.at(j);
+      size_t k = 0;
+      for (size_t l = 0; l < Fj.size(); ++l) {
+        if (!subsumes(constraint, Fj[l])) {
+          Fj[k++] = Fj[l];
+        }
+      }
+      Fj.resize(k);
+    }
+  }
+
+  assert(i > 0);  // there's a special case for frame 0
+
+  constrain_frame_label(i, constraint);
+  frames_.at(i).push_back(constraint);
+}
+
+// -----------------------------------------------------------------
+// Require the capability to reset solver
+// -----------------------------------------------------------------
+
+void SygusPdr::reset_solver()
```

Comment:
This looks the same as the base class implementation so it shouldn't be overridden. If you need to do extra work in `reset_solver` I think the best thing to do is override, and use `super::reset_solver` to do the basic parts.

Reply:
btw, I'll refactor that soon so that adding the label constraints is separate from resetting the solver. I'll implement that tomorrow and let you know.